merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/ActionTree.java

```diff
@@ -0,0 +1,186 @@
+  }
+
+  private static Task<Unit, Unit> spawnAll(final Topic<ActivityInstanceId> activityTopic, final List<Pair<Duration, ActionTree>> actions) {
+    return (scheduler, unit) -> {
+      for (final var entry : actions) {
+        final var startOffset = entry.getLeft();
+        final var action = entry.getRight();
+
+        scheduler.spawn(
+            $$ -> Task
+                .spawning(action.toTask(activityTopic))
+                .butFirst(Task.delaying(startOffset)),
+            Unit.UNIT);
+      }
+
+      return TaskStatus.completed(Unit.UNIT);
+    };
+  }
+
+  /**
+   * Recursively re-schedule each action relative to the earliest resolvable point in the plan.
+   *
+   * <p> The following transformations are repeatedly made to the plan until no further transformation
+   * is possible. </p>
+   *
+   * <ul>
+   *   <li> If an action (the "child") is anchored to the start of another action (its "parent"), and its parent itself
+   *   has a parent, then the child is re-anchored to its parent's parent, and the offset of its parent is added to
+   *   its own offset. </li>
+   *
+   *   <li> If an action is anchored to the end of a plan or event, it is re-anchored to the start of that plan
+   *   or event, and the duration of the plan or event is added to the action's offset. </li>
+   * </ul>
+   *
+   * <p> Neither of these transformations affects the absolute time at which an action is scheduled. Instead, these
+   * transformations cause an action to become anchored the earliest point at which we know its precise start time.
+   * Notably, any actions anchored to the end of an activity cannot be migrated any earlier, as we do not know the
+   * duration of an activity until it has actually terminated. </p>
+   *
+   * @return
+   *   A new, equivalent plan with all actions re-anchored to the first point at which its absolute start time
+   *   becomes known.
+   */
+  public ActionTree resolveOffsets() {
+    final var startAnchor = new ArrayList<Pair<Duration, ActionTree>>(this.startAnchor.size());
+
+    // Recursively resolve each start-anchored child, then add it and its own start-anchored children to ours.
+    // But, instead of doing this in two steps, we pass to each child its own offset and the accumulating list of
+    // start-anchored actions, so that it can add itself and any descendants directly.
+    final var self = this.liftInto(startAnchor, Duration.ZERO);
+
+    // `self.startAnchor` should always be empty after lifting, but to be robust against future changes,
+    // there's no harm in accumulating it anyway.
+    startAnchor.addAll(self.startAnchor);
+
+    return new ActionTree(self.action(), startAnchor, self.endAnchor());
+  }
+
+  /**
+   * Lift as many descendant actions as possible into the parent's anchor, and return an ActionTree with all remaining
+   * actions lifted as high as possible.
+   *
+   * @param anchor
+   *   The anchor of the parent which this action tree is scheduled relative to.
```

Comment:
@Mythicaeda mentioned finding this line difficult to grasp.

Since this is a recursive method, I recommend understanding `anchor` in context of one of its call sites. [Lines 173-180](https://github.com/NASA-AMMOS/aerie/pull/531/commits/c96fc885285fb1ae9c1d07579e455bd461f83b0f#diff-fdc43d758da16322322b30bba1f8078d2c7891a41e22ba7a469615473b2130abR173-R180) show where we pass on a different value for `anchors` than what we received. In that block, since we can't resolve through a directive, the new "best place" to move actions to is the current end anchor.